src/rockstor/storageadmin/views/email_client.py

```diff
@@ -19,46 +19,130 @@
 import os
 import re
 import json
+import stat
+import distro
 from rest_framework.response import Response
 from django.db import transaction
-from storageadmin.models import (EmailClient, Appliance)
+from storageadmin.models import EmailClient, Appliance
 from storageadmin.serializers import EmailClientSerializer
 from storageadmin.util import handle_exception
 import rest_framework_custom as rfc
-from system.osi import run_command, gethostname
+from system.osi import run_command, gethostname, replace_line_if_found
 from shutil import move
 from tempfile import mkstemp
 from system.services import systemctl
 from system.email_util import send_test_email, test_smtp_auth
 import logging
+
 logger = logging.getLogger(__name__)
 
 
-POSTMAP = '/usr/sbin/postmap'
-HEADER = '####BEGIN: Rockstor section####'
-FOOTER = '####END: Rockstor section####'
-MAIN_CF = '/etc/postfix/main.cf'
+POSTMAP = "/usr/sbin/postmap"
+HEADER = "####BEGIN: Rockstor section####"
+FOOTER = "####END: Rockstor section####"
+MAIN_CF = "/etc/postfix/main.cf"
+MASTER_CF = "/etc/postfix/master.cf"
+POSTFIX = "/usr/sbin/postfix"
+SYSCONFIG_MAIL = "/etc/sysconfig/mail"
+ROOT_FORWARD = "/root/.forward"
+GENERIC = "/etc/postfix/generic"
+
+# List of our used config options within main.cf; used to remove prior entries
+# to avoid log warnings of "... overriding earlier entry: X"
+MAIN_CF_OPTIONS = [
+    "relayhost",
+    "smtp_use_tls",
+    "smtp_sasl_auth_enable",
+    "smtp_sasl_password_maps",
+    "smtp_tls_CAfile",
+    "smtp_sasl_security_options",
+    "smtp_sasl_tls_security_options",
+    "smtp_generic_maps",
+]
 
 
 def rockstor_postfix_config(fo, smtp_server, port, revert):
-    if (revert is True):
+    if revert is True:
+        return
+    distro_id = distro.id()
```

Comment:
@phillxnet,
A quick comment before I can test this one out: I noticed you are calling `distro.id()` every time here. In line with our discussion on a different PR related to using variables defined at build time, I actually used `distro_id = settings.OS_DISTRO_ID` instead in my recently submitted PR believing it would be slightly faster and/or more efficient.

Would that actually be the case? The difference is probably rather minor but I wanted to know what you think of it.

Reply:
@FroggyFlox Re:

> In line with our discussion on a different PR related to using variables defined at build time, I actually used `distro_id = settings.OS_DISTRO_ID` instead in my recently submitted PR believing it would be slightly faster and/or more efficient.

Yes, good point, I'd actually forgotten about that; however as you say it's a fairly trivial time saver in this case as it's only going to be called once/twice during each email setup, so maybe no more than a handful of times during an installs life time. My original intent with the settings.OS_DISTRO_ID was to save the "Built on ..." / "Uses ..." info and to avoid the repeated calls that are inevitable with it's use in the Web-UI header. I didn't want the Web-UI refresh to be calling the distro library at every refresh interval, though that's only something like every 10 to 30 minutes it's far more pressing a concern than what we have here.

So in short I'd suggest we absolutely use the settings var for frequent calls where we can, i.e. where Leap to Tumbleweed and back don't matter. But my most recent significant use of this library was in the packaging area where we have to account for folks moving their install from Leap to Tumbleweed and back, as they have different repositories. And that was more fresh than the Web-UI header work. Plus the header then denotes what the rpm was built on, or that was the intention, giving added debug value to a screen shot.

I say we leave this as is for now as it's 'mostly harmless' and have more of a talk on the best way to go in the future. In most cases the Django settings is probably best, so in turn I say we also leave that 'as-is' in your current pull request. But we should very definitely avoid frequent calls to distro if we can, and the setting var gives us that option. Plus there may be a related memory trade-off in pulling in the settings but I'm just not up enough of Django workings to know. So swings and roundabouts and I say we go with what we have both used thus far but definitely avoid, if possible, repeated calls. You samba config calls are also likely to only be used once per config, which is not that often. And if there turns out to be a difference, say between Leap / Tumbleweed and folks move there system under us then we will miss it, at least until a new rpm is pulled down. But that is unlikely currently. So yes lets chat on this but get these two pull requests sorted if we can in the mean time.

Nicely spotted though, and brings to the fore the question of how we use this. It's all a bit messy really but I couldn't thing of another way around it while we are in this transition. We might consider ripping out all uses of this now oft used library once we are done with the CentOS target all together.

Incidentally in the recent package work I also tried to call a minimum number of times by calling early and passing the result to the multiple child processes. I had though of doing that here but for a single call per email setup it seemed like unnecessary complexity. I should probably have noted this in the pr.

Thanks for bringing this up.

Reply:
Thanks a lot for the very detailed answer!